packages/ui-extensions-sdk/src/utils/security.ts

```diff
@@ -0,0 +1,113 @@
+import { RESOURCE_BATCH_INTERVAL } from '../constants';
+import { LoadedResourceMetaDataBatch } from '../types';
+
+import { setImmediateInterval } from './utils';
+
+export type LoadedResourceIds = Set<string>;
+
+const isPerformanceObjectSupported = () => {
+    return window.performance !== undefined && 'getEntries' in performance;
+};
+
+const isResourceTimingEntry = (
+    entry: PerformanceEntry
+): entry is PerformanceResourceTiming => {
+    return entry.entryType === 'resource';
+};
+
+const isRequestKind = (timing: PerformanceResourceTiming) => {
+    return (
+        timing.initiatorType === 'xmlhttprequest' ||
+        timing.initiatorType === 'fetch'
+    );
+};
+
+export const collectResourceUsage = (
+    previouslyLoadedResources: LoadedResourceIds
+): [LoadedResourceMetaDataBatch, LoadedResourceIds] => {
+    // Use resource timing api to to retrieve application loaded resources
+    const snapshot = performance.getEntriesByType('resource');
+
+    const resources = snapshot
+        .filter(isResourceTimingEntry)
+        // to avoid collecting the same resource twice, we use
+        // the Set difference: snapshot - previousSnapshot
+        .filter(r => !previouslyLoadedResources.has(r.name))
+        .map(r => {
+            // safe because r.name represents the resolved URL of the requested resource.
+            const urlHostname = new URL(r.name).hostname;
+            const ts = performance.timeOrigin;
+            let url;
+            if (!isRequestKind(r)) {
+                // for privacy reason we do not capture full URLs of network calls
+                url = r.name;
+            }
+            return {
+                startTimeTs: ts + r.startTime,
+                secureConnectionStartTs: ts + r.secureConnectionStart,
+                url,
+                urlHostname,
+                initiatorType: r.initiatorType,
+                nextHopProtocol: r.nextHopProtocol,
+                duration: r.duration,
+                decodedBodySize: r.decodedBodySize
```

Comment:
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**❗ warning:** ‏FYI, most of the timing / size attributes won't be available for [cross origin resources](https://www.w3.org/TR/resource-timing-2/#sec-cross-origin-resources)

Reply:
Thanks, that should be fine, it just means incomplete data in some cases